neuralmonkey/trainers/objective.py

```diff
@@ -0,0 +1,81 @@
+from abc import abstractproperty
+from typing import TypeVar, Union, Tuple, List, Optional, Generic
+import tensorflow as tf
+from typeguard import check_argument_types
+
+from neuralmonkey.decorators import tensor
+from neuralmonkey.model.model_part import GenericModelPart
+
+# pylint: disable=invalid-name
+ObjectiveWeight = Union[tf.Tensor, float, None]
+Gradients = List[Tuple[tf.Tensor, tf.Variable]]
+MP = TypeVar("MP", bound=GenericModelPart)
+# pylint: enable=invalid-name
+
+
+class Objective(Generic[MP]):
```

Comment:
Proč je to generický? (A kde si pak používá ten atribut dekodér.) Teoreticky ten dekodér taky ani nemusí bejt jeden, viz jak jsme nedávno psali tu hinge loss, co brala dva statefúly.

Reply:
Generic trainer si nastavuje dependencies podle dekodérů všech svých objectives.
Takže decoder musí viset na base třídě a tím pádem aby fungovalo pořádně mypy, musí bejt generická podle typu dekodéru. Protože kdyby byl v definici objective dekodér jako `GenericModelPart`, tak by mypy řvalo u všech poděděných objectives s konkrétnějším dekodérem, že to takhle nejde.